src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj

```diff
@@ -13,7 +13,7 @@
     <PublishDir>$(RuntimeBinDir)ilc-published/</PublishDir>
     <PublishTrimmed>true</PublishTrimmed>
     <!-- Don't R2R on ARM64 machines because ARM64 crossgen2 that comes with .NET SDK <= 7.0 Preview 7 crashes.-->
-    <PublishReadyToRun Condition="'$(BuildArchitecture)' != 'arm64'">true</PublishReadyToRun>
+    <PublishReadyToRun Condition="'$(BuildArchitecture)' != 'arm64' AND '$(DotNetBuildFromSource)' != 'true' ">true</PublishReadyToRun>
```

Comment:
This change works around:
```
error NETSDK1095: Optimizing assemblies for performance is not supported for the selected target platform or architecture. Please verify you are using a supported runtime identifier, or set the PublishReadyToRun property to false.
```
I don't know why the SDK thinks it doesn't support the platform?

Reply:
This looks weird, I agree. One of these conditions must be false to trigger the sdk error: https://github.com/dotnet/sdk/blob/b99ebbe963623467a37c327e2544af1381b1d2c3/src/Tasks/Microsoft.NET.Build.Tasks/ResolveReadyToRunCompilers.cs#L112-L115.

Reply:
I wonder if it's the straight check ` _targetPlatform != hostPlatform`. It's gonna compare non-portable and portable platforms.

Reply:
This is the Task that fails with its parameters:
```
  13:47:14.530    19>Target "ResolveReadyToRunCompilers: (TargetId:2247)" in file "/home/tmds/sdk/sdk/7.0.100-rtm.22478.1/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets" from project "/home/tmds/tarball41/src/runtime/artifacts/source-build/self/src/src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj" (target "_PrepareForReadyToRunCompilation" depends on it):
                     Using "ResolveReadyToRunCompilers" task from assembly "/home/tmds/sdk/sdk/7.0.100-rtm.22478.1/Sdks/Microsoft.NET.Sdk/targets/../tools/net7.0/Microsoft.NET.Build.Tasks.dll".
                     Task "ResolveReadyToRunCompilers" (TaskId:1425)
                       Task Parameter:ReadyToRunUseCrossgen2=True (TaskId:1425)
                       Task Parameter:
                           TargetingPacks=
                               Microsoft.NETCore.App
                                       NuGetPackageId=Microsoft.NETCore.App.Ref
                                       NuGetPackageVersion=7.0.0-rtm.22465.3
                                       PackageConflictPreferredPackages=Microsoft.NETCore.App.Ref;Microsoft.NETCore.App.Runtime.fedora.36-x64
                                       PackageDirectory=/home/tmds/sdk/packs/Microsoft.NETCore.App.Ref/7.0.0-rtm.22465.3
                                       Path=/home/tmds/sdk/packs/Microsoft.NETCore.App.Ref/7.0.0-rtm.22465.3
                                       RuntimeFrameworkName=Microsoft.NETCore.App
                                       RuntimePackRuntimeIdentifiers=fedora.36-x64
                                       TargetFramework=net7.0
                                       TargetingPackFormat= (TaskId:1425)
                       Task Parameter:NETCoreSdkRuntimeIdentifier=fedora.36-x64 (TaskId:1425)
                       Task Parameter:
                           Crossgen2Packs=
                               Microsoft.NETCore.App.Crossgen2.fedora.36-x64
                                       NuGetPackageId=Microsoft.NETCore.App.Crossgen2.fedora.36-x64
                                       NuGetPackageVersion=7.0.0-rtm.22465.3
                                       OriginalItemSpec=Microsoft.NETCore.App.Crossgen2.fedora.36-x64
                                       PackageDirectory=/home/tmds/tarball41/src/runtime/artifacts/source-build/self/package-cache/microsoft.netcore.app.crossgen2.fedora.36-x64/7.0.0-rtm.22465.3 (TaskId:1425)
                       Task Parameter:
                           RuntimePacks=
                               Microsoft.NETCore.App.Runtime.fedora.36-x64
                                       FrameworkName=Microsoft.NETCore.App
                                       IsTrimmable=
                                       NuGetPackageId=Microsoft.NETCore.App.Runtime.fedora.36-x64
                                       NuGetPackageVersion=7.0.0-rtm.22465.3
                                       OriginalItemSpec=Microsoft.NETCore.App.Runtime.fedora.36-x64
                                       PackageDirectory=/home/tmds/tarball41/src/runtime/artifacts/source-build/self/package-cache/microsoft.netcore.app.runtime.fedora.36-x64/7.0.0-rtm.22465.3
                                       RuntimeIdentifier=fedora.36-x64 (TaskId:1425)
                       Task Parameter:PerfmapFormatVersion=1 (TaskId:1425)
                       Task Parameter:RuntimeGraphPath=/home/tmds/tarball41/src/runtime/artifacts/source-build/self/src/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json (TaskId:1425)
  13:47:14.535    19>/home/tmds/sdk/sdk/7.0.100-rtm.22478.1/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(299,5): error NETSDK1095: Optimizing assemblies for performance is not supported for the selected target platform or architecture. Please verify you are using a supported runtime identifier, or set the PublishReadyToRun property to false. [/home/tmds/tarball41/src/runtime/artifacts/source-build/self/src/src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj]
```

> I wonder if it's the straight check _targetPlatform != hostPlatform. It's gonna compare non-portable and portable platforms.

```
            // Get the best RID for the host machine, which will be used to validate that we can run crossgen for the target platform and architecture
            _hostRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(
                runtimeGraph,
                NETCoreSdkRuntimeIdentifier,
                supportedRIDsList,
                out _);
```
is probably going to be `fedora.36-x64`.

```
            _targetRuntimeIdentifier = _runtimePack?.GetMetadata(MetadataKeys.RuntimeIdentifier);
```
what does this read?

Maybe this gets fixed as part of resolving https://github.com/dotnet/runtime/issues/74721.

Reply:
> what does this read?

This should also return `fedora.36-x64`( from: `RuntimeIdentifier=fedora.36-x64`).

When determining `_hostRunttimeIdentifier`, `NETCoreSdkRuntimeIdentifier` is `fedora.36-x64` and `supportedRIDsList` is `fedora.36-x64`.

So it looks like `_targetPlatform` and `hostPlatform` will be the same.

Skipping `PublishReadyToRun` here when building from source is fine, though it would be nice if we understand the root cause.

Reply:
@ViktorHofer @hoyosjs should we be looking at `private bool ValidateCrossgen2Support()`?

https://github.com/dotnet/sdk/blob/b99ebbe963623467a37c327e2544af1381b1d2c3/src/Tasks/Microsoft.NET.Build.Tasks/ResolveReadyToRunCompilers.cs#L145-L162

It looks like `targetOS` will be `null` when `_targetRuntimeIdentifier` is a non-portable rid like `fedora.36-x64`.

Reply:
I've created https://github.com/dotnet/sdk/issues/28312 for this.

Reply:
This is fixed by https://github.com/dotnet/sdk/pull/28380, and makes this change unnecessary.